susemanager-utils/susemanager-sls/modules/roster/uyuni.py

```diff
@@ -0,0 +1,309 @@
+            % (
+                SSH_KEY_PATH if i == 0 else PROXY_SSH_PUSH_KEY,
+                PROXY_SSH_PUSH_USER,
+                "-W %s:%s" % (minion_id, ssh_push_port)
+                if not tunnel and i == len(proxies) - 1
+                else "",
+                proxy,
+            )
+        )
+        i += 1
+    if tunnel:
+        proxyCommand += (
+            "/usr/bin/ssh -i {pushKey} -o StrictHostKeyChecking=no "
+            "-o User={user} -R {pushPort}:{proxy}:{sslPort} {minion} "
+            "ssh -i {ownKey} -W {minion}:{sshPort} "
+            "-o StrictHostKeyChecking=no -o User={user} {minion}".format(
+                pushKey=PROXY_SSH_PUSH_KEY,
+                user=user,
+                pushPort=SSH_PUSH_PORT_HTTPS,
+                proxy=proxies[len(proxies) - 1],
+                sslPort=SSL_PORT,
+                minion=minion_id,
+                ownKey="{}{}".format(
+                    "/root" if user == "root" else "/home/{}".format(user),
+                    "/.ssh/mgr_own_id",
+                ),
+                sshPort=ssh_push_port,
+            )
+        )
+    proxyCommand += "'"
+
+    return [proxyCommand]
+
+
+def _getSSHMinion(
+    minion_id=None, proxies=[], tunnel=False, ssh_push_port=SSH_PUSH_PORT
+):
+    user = SSH_PUSH_SUDO_USER if SSH_PUSH_SUDO_USER else "root"
+    minion = {
+        "host": minion_id,
+        "user": user,
+        "port": ssh_push_port,
+        "timeout": SALT_SSH_CONNECT_TIMEOUT,
+    }
+    if tunnel:
+        minion.update({"minion_opts": {"master": minion_id}})
+    if proxies:
+        minion.update(
+            {
+                "ssh_options": _getSSHOptions(
+                    minion_id=minion_id,
+                    proxies=proxies,
+                    tunnel=tunnel,
+                    user=user,
+                    ssh_push_port=ssh_push_port,
+                )
+            }
+        )
+    elif tunnel:
+        minion.update(
+            {
+                "remote_port_forwards": "%d:%s:%d"
+                % (SSH_PUSH_PORT_HTTPS, COBBLER_HOST, SSL_PORT)
+            }
+        )
+
+    return minion
+
+
+def targets(tgt, tgt_type="glob", **kwargs):
+    """
+    Return the targets from the Uyuni DB
+    """
+
+    ret = {}
+
+    cache_data = cache.fetch("roster/uyuni", "minions")
+    cache_fp = cache_data.get("fp", None)
+    query = """
+        SELECT FORMAT('%s|%s|%s|%s',
+                      EXTRACT(EPOCH FROM MAX(S.modified)),
+                      COUNT(S.id),
+                      EXTRACT(EPOCH FROM MAX(SP.modified)),
+                      COUNT(SP.proxy_server_id)
+               ) AS fp
+               FROM rhnServer AS S
+               INNER JOIN suseServerContactMethod AS SSCM ON
+                     (SSCM.id=S.contact_method_id)
+               INNER JOIN suseMinionInfo AS SMI ON
+                     (SMI.server_id=S.id)
+               LEFT JOIN rhnServerPath AS SP ON
```

Comment:
since you are using left join, the rhnServerPath may not have a matching entry. This will lead to empty data in return.
I'm assuming that is what you want, and that case represents that minion is connected directly to a server, without any proxy.

Reply:
yes, you are right. `rhnServerPath` could contain no records for the minion, but I selecting all the records in one query and there is additional logic in python to build proper structure.